torch/onnx/_internal/fx/passes/fx_to_onnxscript.py

```diff
@@ -0,0 +1,510 @@
+    tracer = graph_building.TorchScriptTracingEvaluator(onnxscript_graph)
+
+    # In the following loop, a TorchScript graph is created to
+    # represent the input FX graph with ONNX symbols (e.g., onnx::add).
+    # To connect the values to nodes in the TorchScript graph, we maintian
+    # fx_name_to_onnxscipt_value. Basically, we want to translate
+    #   fx_tensor_x (type: torch.fx.Node) -> fx_node_1 -> fx_tensor_y (type: torch.fx.Node)
+    # to
+    #   fx_name_to_onnxscipt_value[fx_tensor_x.name] -> onnx_node_1 -> fx_name_to_onnxscipt_value[fx_tensor_y.name]
+    fx_name_to_onnxscipt_value: Dict[
+        str, Union[torch._C.Value, Tuple[torch._C.Value, ...]]
+    ] = {}
+    # Similar to fx_name_to_onnxscipt_value, we need a mapping fo real tensors (usually tensor parameters
+    # in nn.Module). Note that TorchScript's cannot store real tensors; TorchScript values are all
+    # symbolic. This is passed into ONNX ModelProto as the initializers.
+    onnxscript_value_name_to_real_tensor: Dict[
+        str, Union[torch.Tensor, Tuple[torch._C.Value, ...]]
+    ] = {}
+    for node in fx_module_with_metadata.graph.nodes:
+        _export_fx_node_to_onnxscript(
+            node,
+            onnxscript_graph,
+            fx_name_to_onnxscipt_value,
+            onnxscript_value_name_to_real_tensor,
+            tracer,
+            fx_module_with_metadata,
+            options,
+        )
+
+    # Apply TorchScript's type promotion code.
+    # Ideally, we should implement our type promotion but
+    # to save time, we just reuse.
+    onnxscript_graph.apply(
+        torch._C._jit_pass_onnx_scalar_type_analysis,
+        lowprecision_cast=True,
+        opset_version=options.opset_version,
+    )
+
+    return onnxscript_graph, onnxscript_value_name_to_real_tensor
+
+
+@_beartype.beartype
+def _validate_op_between_ort_torch(
+    node: torch.fx.Node,
+    symbolic_fn: Union[onnxscript.OnnxFunction, Callable],
+    torch_args: tuple,
+    torch_kwargs: dict,
+):
+    """Validate the op between ONNX Runtime and PyTorch."""
+    # op-level validation
+    # Symbolic_fn should have the same output as node.target (torch ops)
+    # trace_only function is regular python function
+    function_name = (
+        symbolic_fn.name
+        if isinstance(symbolic_fn, onnxscript.OnnxFunction)
+        else symbolic_fn.__name__
+    )
+    try:
+        with evaluator.default_as(evaluator.ort_evaluator):
+            expected_outputs = node.target(*torch_args, **torch_kwargs)  # type: ignore[operator]
+            # TODO(titaiwang): Expose _convert_tensor_to_numpy and _convert_kwargs_for_onnx?
+            input_onnx = [
+                onnx_proto_utils._convert_tensor_to_numpy(x) for x in torch_args
+            ]
+            kwargs_onnx = _filter_incompatible_and_dtype_convert_kwargs(torch_kwargs)
+            ort_outputs = symbolic_fn(*input_onnx, **kwargs_onnx)
+
+            # TODO: add pytree structure comparison.
```

Comment:
Impossible to compare pytree structure. FX allows recursive collections such as `List[Dict[List[.....]`.

Reply:
Could you elaborate what does `recursive collections` mean? Is it just nested containers, or are you suggesting it's possible to have cycles?

And could you elaborate why such structure makes it impossible to compare?

Reply:
Reopen the function related issue that we could have more talk about this: https://github.com/microsoft/onnx-script/issues/393